Round-trip OTLP-OTAP-OTLP validation framework

jmacd opened the pull request:
The goal is to test the OTAP receiver logic originally contributed in otel-arrow-rust using the pdata library to build OTLP objects for readable tests.

Part of #338 
Part of #353

utpilla commented on rust/otel-arrow-rust/src/pdata/mod.rs:
```diff
@@ -23,6 +23,18 @@ pub mod otlp;
+type Error = &'static str;
+
+impl<'a> TraceID {
+    pub fn try_new(value: &[u8]) -> Result<TraceID, Error> {
```
You could enforce the length check at compile time by updating the method argument to an array or an array reference.

```rust
impl TraceID {
    pub fn new(value: &[u8; 16]) -> TraceID {
        TraceID(*value)
    }
}
```

jmacd commented:
WIP: @v0y4g3r I have built here a framework for testing the OTAP to OTLP conversion in metrics using the Golang `otelarrow` exporter. 😀 

See [b796f80](https://github.com/open-telemetry/otel-arrow/pull/346/commits/b796f80c05481c5daf367f0713cbe040f838a191) I was able to find and fix a small issue, similar to one you had shown earlier. Please take a look.

lquerel commented on rust/otel-arrow-rust/src/validation/service_type.rs:
```diff
@@ -0,0 +1,163 @@
+    fn protocol() -> &'static str;
+
+    /// Create a new client for this service
+    async fn connect_client(endpoint: String) -> Result<Self::Client, tonic::transport::Error>;
+
+    /// Send data through the client
+    async fn send_data(
+        client: &mut Self::Client,
+        request: Self::Request,
+    ) -> Result<Self::Response, tonic::Status>;
```
Whenever possible, I try to avoid exposing types from dependencies in the public API. This minimizes the risk of introducing breaking changes in our APIs if one of our dependencies changes.

jmacd replied:
I wasn't thinking of this as a public API-- this would be internal and internal validation package used for checking the Golang and Rust implementations at first. Otherwise, I am not sure how to isolate -- is it the `Client` field? How should I fix it?

lquerel reviewed (approved):
I made a few general comments to make future integration into the main project easier. To avoid unpleasant surprises, I think you should apply the same Clippy rules in this project, otherwise, the integration might be painful :-)

v0y4g3r commented on rust/otel-arrow-rust/src/validation/collector.rs:
```diff
@@ -0,0 +1,378 @@
+});
+
+/// TimeoutError represents an error when a receiver operation times out
+#[derive(Debug)]
+pub struct TimeoutError {
+    pub duration: Duration,
+}
```
> I suggest to use thiserror to define structured errors.

[thiserror](https://docs.rs/thiserror/latest/thiserror/) and [snafu](https://docs.rs/snafu/latest/snafu/) both are good choices, for you reference we've introduced snafu here:

https://github.com/open-telemetry/otel-arrow/blob/913bdf06fb76527ac886572a7ad3aecf7e2e9b44/rust/otel-arrow-rust/src/error.rs#L23-L40

jmacd replied:
I don't want to force a debate over snafu vs. thiserror at this time. :-) 
I will remove the new error in favor of snafu, which is already in this crate.

v0y4g3r commented:
> WIP: @v0y4g3r I have built here a framework for testing the OTAP to OTLP conversion in metrics using the Golang `otelarrow` exporter. 😀
> 
> See [b796f80](https://github.com/open-telemetry/otel-arrow/pull/346/commits/b796f80c05481c5daf367f0713cbe040f838a191) I was able to find and fix a small issue, similar to one you had shown earlier. Please take a look.

That's great. 

> WIP: @v0y4g3r I have built here a framework for testing the OTAP to OTLP conversion in metrics using the Golang `otelarrow` exporter. 😀
> 
> See [b796f80](https://github.com/open-telemetry/otel-arrow/pull/346/commits/b796f80c05481c5daf367f0713cbe040f838a191) I was able to find and fix a small issue, similar to one you had shown earlier. Please take a look.

Thanks Josh, the test suite roughly looks good to me. It will help a lot to find the difference between implementations.

jmacd commented on rust/otel-arrow-rust/src/validation/collector.rs:
```diff
@@ -0,0 +1,378 @@
+            .stderr(Stdio::piped())
+            .spawn()
+            .map_err(|e| format!("Failed to start collector process: {}", e))?;
+
```
Ahh, ok. I wasn't are of that, will do.

jmacd commented on rust/otel-arrow-rust/src/validation/collector.rs:
```diff
@@ -0,0 +1,378 @@
+    F: FnOnce(TestContext<I, O>) -> T,
+    T: std::future::Future<Output = (TestContext<I, O>, Result<(), Box<dyn std::error::Error>>)>,
+{
+    // Generate random ports in the high u16 range to avoid conflicts
+    let random_value = rand::random::<u16>();
+    let receiver_port = 40000 + (random_value % 25000);
```
I'll add a comment. The problem I'm working around is that this is the port a collector will open, we can't open it in this process. We can use the `:0` port specifier to have the Golang process allocate a port, but it doesn't print the port it allocates at normal logging level. I could parse the gRPC debug logs, maybe, to get this info, but this works somewhat. I would probably have the test detect failure to start because of a port in use and have it retry at the test framework level to work around.

jmacd reviewed (commented):
I will work on clippy; I'll just copy what's being done in the other crates and see!

As for the actual validation task here, I reverted my changes to src/otlp/metric.rs, leaving the original discovery as a failure test case. For the OTLP->OTAP->OTLP test here:

```
    #[tokio::test]
    async fn test_otap_metrics_single_request() {
        run_single_round_trip_test::<OTLPMetricsInputType, OTAPMetricsOutputType, _>(
            testdata::metrics::create_single_request,
	    // This test expects a specific error due to disagreements
	    // between the Rust and Golang implementations about OTAP
	    // metrics encoding.  See issue 
            Some("ColumnDataTypeMismatch"),
        )
        .await;
    }
```
we are testing a failure. I found several problems which made me less confident in my approach, so I will try to fix those issues after addressing the feedback in this PR. They are:

- units: needs string dictionary accessor
- resource_arrays: can be None when there are only empty resources
- scope_arrays: can be None when there are only empty scopes
- metric_arrays: some kind of column-mapping problems appear, like it expects metric_id to be uint16 but actually finds a uint32 column (which I suspect is something else); I need to learn more detail about OTAP

jmacd commented:
Reviewers, I believe I've addressed all the feedback.

- Structured errors throughout (using `snafu` because of the precedent here, I will form an opinion about it vs `thiserror`)
- Convert to tokio::process::Command, remove use of threads and sync/async adapters
- Simplify and improve readability.

jmacd commented:
I attempted to fix the otel-arrow-rust tests. The `make otelarrowcol` step needs to be run in the CI/CD workflow to build an otelarrowcol binary for the Rust tests (or we could use a docker image, etc). I think it will be best to merge this and diagnose the failures. Tests pass locally.

jmacd commented:
I tried to address the formatting disagreement (between prost/tonic and rustfmt), which appears to be about editions. I tried to resolve that as well, but could not (see the new rustfmt.toml file).

FYI all `clippy` tests pass, edition is set to 2024, and tonic/prost are set to matching versions.
